{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41c684e2-1029-4c47-89b9-4d5d5ec33179",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import plotnine as pn\n",
    "import pyranges as pr\n",
    "from scipy.stats import linregress\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from pyarrow.parquet import ParquetFile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "223254ed-d2ef-47a9-b9b8-ab9a07081e1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_file = \"/s/genomes/Gencode/Gencode_human/release_34/GRCh37_mapping/gencode.v34lift37.annotation.gtf.gz\"\n",
    "genome_annotation = pr.read_gtf(gtf_file)\n",
    "gene_annotation = genome_annotation[genome_annotation.Feature==\"gene\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c65075f-dc30-4de8-a72c-f05975a5e398",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Read parquet files and convert to pandas DataFrame\n",
    "dt = pd.read_parquet('/s/project/geno2pheno/predictions/bayesian/best_model_pred/v108cov_deepRVAT_genes_extended.pq')\n",
    "dt_ols = pd.read_parquet('/s/project/geno2pheno/predictions/bayesian/best_model_pred/ols_filteredv3_deepRVAT_genes_extended.pq')\n",
    "# dt_flat = pd.read_parquet('/s/project/geno2pheno/predictions/bayesian/v107cov_flat_deepRVAT_genes_extended.pq')\n",
    "\n",
    "# Concatenate DataFrames\n",
    "dt = pd.concat([dt[['model', 'trait', 'phenocode', 'gene_id', 'neglog_pval', 'pd', 'significant', 'replicated', 'beta']], \n",
    "                dt_ols[['model', 'trait', 'phenocode', 'gene_id', 'neglog_pval', 'significant', 'replicated', 'beta']]], \n",
    "                ignore_index=True)\n",
    "\n",
    "# Load gene names\n",
    "gene_names = pd.read_csv('/s/project/geno2pheno/data/hgnc2ensg.tsv', sep='\\t')[['Ensembl gene ID', 'Approved symbol']].drop_duplicates().rename(columns={'Ensembl gene ID':'gene_id', 'Approved symbol':'gene_name'})\n",
    "\n",
    "# Merge table with gene names\n",
    "dt = dt.merge(gene_names, on='gene_id')\n",
    "dt['significant'] = np.where(dt['pd']>0.999, True, False)\n",
    "dt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ee100f5-cd8a-4774-9742-767de418c127",
   "metadata": {},
   "outputs": [],
   "source": [
    "trait_gene_dict = {\n",
    "    'Lymphocyte_percentage':'ARHGAP15',   # 30180\n",
    "    'Platelet_crit':'RHOF',               # 30090\n",
    "}\n",
    "\n",
    "for tr in trait_gene_dict:\n",
    "    print(tr, trait_gene_dict[tr])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52859382-1ee3-4819-9373-66dead96d68f",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt_genes = pd.concat([dt[(dt.trait==tr) & (dt.gene_name == trait_gene_dict[tr]) & (dt.model=='omics_pops_bayesian_v108cov_deepRVAT')] for tr in trait_gene_dict])\n",
    "plt_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9532810f-2dca-45cf-9d09-a1d305d5c2fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "nom_pval_dt = pd.concat([dt[(dt.trait==tr) & (dt.gene_name == trait_gene_dict[tr]) & (dt.model=='ols_deepRVAT_cov')] for tr in trait_gene_dict])\n",
    "nom_pval_dt['nom_pval'] = np.exp(-nom_pval_dt['neglog_pval'])\n",
    "# nom_pval_dt\n",
    "plt_genes = pd.merge(plt_genes, nom_pval_dt[['gene_id', 'nom_pval']], on='gene_id')\n",
    "plt_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b0ed04b-c3d7-4638-935a-94ad82e710cd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "trait_cols = [f'{tr}_common_resid' for tr in trait_gene_dict]\n",
    "\n",
    "# 200K UK Biobank\n",
    "# deeprvat_gt = pd.read_parquet(\"/s/project/uk_biobank/processed/derived_datasets/ukbb_wes_200k_DeepRVAT_mean_6_0.parquet\", columns=list(plt_genes.gene_id)).reset_index()\n",
    "# deeprvat_gt = pd.melt(deeprvat_gt, id_vars='individual', value_name='deeprvat')\n",
    "\n",
    "# plof_gt = pd.read_parquet(\"/s/project/uk_biobank/processed/derived_datasets/ukbb_wes_200k_ensembl_loftee_plof_binarized_pivot.parquet\", columns=list(plt_genes.gene_id)).reset_index()\n",
    "# plof_gt = pd.melt(plof_gt, id_vars='individual', value_name='plof')\n",
    "\n",
    "# pheno_dt = pd.read_parquet(\"/s/project/uk_biobank/processed/derived_datasets/ukbb_common_traits_and_covariates_with_split_and_residuals_filteredv3.parquet\", columns=['individual']+trait_cols)\n",
    "# # geno_dt = deeprvat_gt.merge(plof_gt, on=['individual', 'gene']).merge(gene_names, left_on='gene', right_on='gene_id')\n",
    "# plot_dt = pheno_dt.merge(deeprvat_gt.merge(plof_gt, on=['individual', 'gene']), on='individual').merge(gene_names, left_on='gene', right_on='gene_id').drop(columns='gene')\n",
    "# plot_dt\n",
    "\n",
    "# 300K UK Biobank\n",
    "deeprvat_gt = pd.read_parquet(\"/s/project/uk_biobank/processed/derived_datasets/ukbb_wes_300k_DeepRVAT_mean_6_0.parquet\", columns=list(plt_genes.gene_id)).reset_index()\n",
    "deeprvat_gt = pd.melt(deeprvat_gt, id_vars='individual', value_name='deeprvat')\n",
    "\n",
    "plof_gt = pd.read_parquet(\"/s/project/uk_biobank/processed/derived_datasets/ukbb_wes_500k_vep_plof_binarized.parquet\", columns=list(plt_genes.gene_id)).reset_index()\n",
    "plof_gt = pd.melt(plof_gt, id_vars='individual', value_name='plof')\n",
    "\n",
    "pheno_dt = pd.read_parquet(\"/s/project/uk_biobank/processed/derived_datasets/ukbb_common_traits_and_covariates_with_split_and_residuals_300k_filteredv3.parquet\", columns=['individual']+trait_cols)\n",
    "plot_dt = pheno_dt.merge(deeprvat_gt.merge(plof_gt, on=['individual', 'gene']), on='individual').merge(gene_names, left_on='gene', right_on='gene_id').drop(columns='gene')\n",
    "plot_dt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18aadab9-0faf-4567-846a-18957d723c72",
   "metadata": {},
   "outputs": [],
   "source": [
    "#NEW\n",
    "def get_vars_by_gene(var_mac_parquet_file, gene_id, gene_annotation, padding = 0, fillna = True):\n",
    "    var_names = ParquetFile(var_mac_parquet_file).schema.names\n",
    "    split_var_names = pd.Series(var_names[6:]).str.split(\":\", expand=True)\n",
    "    \n",
    "    variants = pr.from_dict({\n",
    "        \"Chromosome\":split_var_names[0].astype(str), \n",
    "        \"Start\":split_var_names[1].astype(int), \n",
    "        \"End\":split_var_names[1].astype(int)+1,\n",
    "        \"var_name\":pd.Series(var_names[6:])\n",
    "    })\n",
    "    \n",
    "    gene_annotation_expanded = gene_annotation.copy()\n",
    "    gene_annotation_expanded.Start = gene_annotation_expanded.Start - padding\n",
    "    gene_annotation_expanded.End = gene_annotation_expanded.End + padding\n",
    "    \n",
    "    included_vars = variants.intersect(gene_annotation_expanded[gene_annotation_expanded.gene_id.str.startswith(gene_id)])\n",
    "    included_vars = included_vars.as_df()[\"var_name\"].to_list() if included_vars else []\n",
    "    mac_df = pd.read_parquet(var_mac_parquet_file, columns = [\"IID\"]+included_vars)\n",
    "    \n",
    "    mac_df[included_vars] = mac_df[included_vars].fillna(mac_df[included_vars].median()).astype(\"Int8\") if fillna else mac_df[included_vars].astype(\"Int8\")\n",
    "    return mac_df\n",
    "\n",
    "\n",
    "def get_common_variant_covariates(gene_id, trait=None, phenocode=None, padding=100_000):\n",
    "    if phenocode is None:\n",
    "        phenocode = phenocode_dict[trait]\n",
    "        \n",
    "    mac_index_vars_parquet = f\"/s/project/deeprvat/clumping/clumping_shubhankar/{phenocode}.0/GWAS_variants_clumped_mac.parquet\" # Eva's results, filtered for Caucasians and more stringent p-val\n",
    "    # mac_index_vars_parquet = f\"/s/project/uk_biobank/processed/clumping/{phenocode}/GWAS_variants_clumped_mac.parquet\"\n",
    "    res_all = get_vars_by_gene(mac_index_vars_parquet, gene_id, gene_annotation, padding = 100_000)\n",
    "    res_all['individual'] = res_all['IID'].str.split('_').str[0]\n",
    "    res = res_all.drop(columns='IID')\n",
    "    res = res[res.individual.isin(plot_dt.individual)]\n",
    "    res.set_index('individual', inplace=True)\n",
    "    res = 2 - res\n",
    "    res.reset_index(inplace=True)\n",
    "    \n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b660d387-76c9-42fa-9f37-fad069cbe12e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "lm_dt = plot_dt.dropna()\n",
    "\n",
    "slopes = []\n",
    "slope_SEs = []\n",
    "pvals = []\n",
    "for index, row in plt_genes.iterrows():\n",
    "    if row.trait in ['Lymphocyte_percentage', 'Platelet_crit', 'glycated_haemoglobin_hba1c', 'Calcium']:\n",
    "        com_covs = get_common_variant_covariates(row.gene_id, phenocode=row.phenocode)\n",
    "        a = pd.merge(com_covs, lm_dt[(lm_dt.gene_name==row.gene_name)][[f'{row.trait}_common_resid', 'individual']], on='individual')\n",
    "        a = a.set_index('individual')\n",
    "    \n",
    "        lm = LinearRegression().fit(a.drop(columns=f'{row.trait}_common_resid'), a[f'{row.trait}_common_resid'])\n",
    "        prediction = lm.predict(a.drop(columns=f'{row.trait}_common_resid'))\n",
    "        residual = a[f'{row.trait}_common_resid'] - prediction\n",
    "        residual = residual.rename(f'{row.trait}_residual')\n",
    "        \n",
    "        lm_dt = pd.merge(lm_dt, residual, on='individual')\n",
    "        slope, intercept, r, p, se = linregress(lm_dt[(lm_dt.gene_name==row.gene_name)]['deeprvat'], lm_dt[(lm_dt.gene_name==row.gene_name)][f'{row.trait}_residual'])\n",
    "\n",
    "    else:\n",
    "        slope, intercept, r, p, se = linregress(lm_dt[(lm_dt.gene_name==row.gene_name)]['deeprvat'], lm_dt[(lm_dt.gene_name==row.gene_name)][f'{row.trait}_common_resid'])\n",
    "        \n",
    "    slopes.append(slope)\n",
    "    slope_SEs.append(se)\n",
    "    pvals.append(p)\n",
    "\n",
    "\n",
    "plt_genes['slope'] = slopes\n",
    "plt_genes['slope_SE'] = slope_SEs\n",
    "plt_genes['pval'] = pvals\n",
    "plt_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "153b516d-1fb6-4f63-8d3f-2e0c3f9f698f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# plt_genes['p_label'] = plt_genes.nom_pval.apply(lambda x: f\"BD p-val: {x:.5f}\")\n",
    "plt_genes['slope_label'] = \"evaluation slope: \" + round(plt_genes[\"slope\"],2).astype(str) + \"Â±\" + round(1.96*plt_genes[\"slope_SE\"],2).astype(str)\n",
    "plt_genes['p_label'] = plt_genes.pval.apply(lambda x: f\"slope p-value: {x:.5f}\")\n",
    "plt_genes['FuncRVP_estimate'] = plt_genes.beta.apply(lambda x: f\"FuncRVP estimate: {x:.2f}\")\n",
    "\n",
    "plt_genes['slope_signif'] = np.where(np.abs(plt_genes[\"slope\"]) > 1.96*plt_genes[\"slope_SE\"], True, False)\n",
    "plt_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81595338-96e4-44b4-8a96-85c8974b89ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "import patchworklib as pw\n",
    "\n",
    "def make_plot(gene_name, trait):\n",
    "    al_plot = (pn.ggplot(plot_dt[plot_dt.gene_name == gene_name], \n",
    "                     pn.aes(x='deeprvat', y=f'{trait}_common_resid')) \n",
    "           + pn.geom_point(pn.aes(color='plof'), alpha=0.5, size=1)\n",
    "           + pn.geom_hline(yintercept = 0, linetype='dashed')\n",
    "           # + pn.geom_smooth(method='lm')\n",
    "           + pn.stat_smooth(method='lm')\n",
    "           + pn.scale_color_manual(name='pLoF', values = [\"darkgrey\", \"red\"])\n",
    "           + pn.ylab(f\"{trait}\\n(adjusted)\")\n",
    "           + pn.xlab(\"DeepRVAT score\")\n",
    "           # + pn.geom_text(pn.aes(x=np.Inf, y=np.Inf, label=\"p_label\"), data=plt_genes[(plt_genes.trait == trait) & (plt_genes.gene_name == gene_name)], size=13)\n",
    "           # + pn.geom_text(pn.aes(x=np.Inf, y=np.Inf, label=\"beta_label\"), data=plt_genes[(plt_genes.trait == trait) & (plt_genes.gene_name == gene_name)], size=13)\n",
    "           # + pn.geom_text(pn.aes(x=0.755, y=3.5, label=\"p_label\"), data=plt_genes[(plt_genes.trait == trait) & (plt_genes.gene_name == gene_name)], size=13)\n",
    "           + pn.geom_text(pn.aes(x=1, y=4, label=\"FuncRVP_estimate\"), data=plt_genes[(plt_genes.trait == trait) & (plt_genes.gene_name == gene_name)], size=14, ha = \"right\")\n",
    "           + pn.geom_text(pn.aes(x=1, y=3.7, label=\"slope_label\"), data=plt_genes[(plt_genes.trait == trait) & (plt_genes.gene_name == gene_name)], size=14, ha = \"right\")\n",
    "           # + pn.geom_text(pn.aes(x=1, y=3.4, label=\"p_label\"), data=plt_genes[(plt_genes.trait == trait) & (plt_genes.gene_name == gene_name)], size=14, ha = \"right\")\n",
    "           + pn.ggtitle(f\"{gene_name}\")\n",
    "           + pn.ylim([-4,4])\n",
    "           + pn.theme_classic()           \n",
    "           + pn.theme(text=pn.element_text(size=15),\n",
    "                      plot_title = pn.element_text(hjust = 0.5),\n",
    "                      legend_position='bottom'))\n",
    "\n",
    "    return al_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3c64254-1308-4d99-8ee6-0198dea7898d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "plot_vec = []\n",
    "# for tr in trait_gene_dict:\n",
    "for tr in ['Lymphocyte_percentage', 'Platelet_crit']:\n",
    "    plot_vec.append(pw.load_ggplot(make_plot(gene_name=trait_gene_dict[tr], trait=tr), figsize=(3,6)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "519f6ecc-3982-4bec-9c96-b5d97eef5cdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "# g1234 = (plot_vec[0]|plot_vec[1])/(plot_vec[2]|plot_vec[3])\n",
    "# g1234 = plot_vec[0]|plot_vec[1]|plot_vec[2]|plot_vec[3]\n",
    "g1234 = plot_vec[0]|plot_vec[1]\n",
    "g1234"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd1f03fb-52ac-4b4e-acfa-e83b30efeadd",
   "metadata": {},
   "outputs": [],
   "source": [
    "g1234.savefig('/s/project/geno2pheno/figures/figure_6_new.png', dpi=96)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10d6e636-ea8d-4b58-a0f7-ef849fef0166",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95b1bf42-75aa-45e4-9db7-061473bb3d30",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c307d1c-4ba5-485d-b033-5ee1c7df1331",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65a5226e-a784-467b-b710-11751a251178",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36534346-8179-4c5b-92ae-a5695589a8f2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2d522a8-38f8-4cc4-bd1b-78fb4f73cb5f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0440a97-339c-4bbd-81dd-c8ca57afdf57",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_id = \"ENSG00000139725\"\n",
    "phenocode = \"30090\"\n",
    "trait = \"Platelet_crit\"\n",
    "com_covs = get_common_variant_covariates(gene_id, phenocode=phenocode)\n",
    "# com_covs\n",
    "\n",
    "lm_dt = plot_dt.dropna()\n",
    "\n",
    "a = pd.merge(com_covs, lm_dt[(lm_dt.gene_id==gene_id)][[f'{trait}_common_resid', 'individual']], on='individual')\n",
    "a = a.set_index('individual')\n",
    "\n",
    "lm = LinearRegression().fit(a.drop(columns=f'{trait}_common_resid'), a[f'{trait}_common_resid'])\n",
    "prediction = lm.predict(a.drop(columns=f'{trait}_common_resid'))\n",
    "residual = a[f'{trait}_common_resid'] - prediction\n",
    "residual = residual.rename(f'{trait}_residual')\n",
    "residual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3991bd5-111a-4be5-86ad-b5d3aab359fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "a[f'{trait}_common_resid'].hist(bins=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "779212ac-21aa-4706-a253-31a4d70a5612",
   "metadata": {},
   "outputs": [],
   "source": [
    "residual.hist(bins=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc45e440-e8ec-4ad2-aed4-46cef7181fbe",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:anaconda-londhe-geno2pheno]",
   "language": "python",
   "name": "conda-env-anaconda-londhe-geno2pheno-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
